tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc

```diff
@@ -882,6 +922,141 @@ LogicalResult reverseBinaryOpsIfBeneficial(Block* block, bool& changed) {
+      bool needReverse = false;
+      auto& transposedUsers = userMap[curOp];
+      LLVM_DEBUG(llvm::dbgs() << "\ttransposedUsers list of curOp ("
+                              << transposedUsers.size() << "):\n");
+      for (auto& transposedUser : transposedUsers) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "\t\t transposedUser: " << *transposedUser << "\n");
+      }
+      for (auto user : curOp->getResult(0).getUsers()) {
+        LLVM_DEBUG(llvm::dbgs() << "\tuser of curOp: " << *user << "\n");
+        if (isa<tensor::DimOp, shape::ShapeOfOp, mhlo::TransposeOp>(user))
+          continue;
+        if (transposedUsers.find(user) != transposedUsers.end()) continue;
+        needReverse = true;
+      }
+      if (!needReverse) {
+        ++numTransposeEliminated;
+        LLVM_DEBUG(llvm::dbgs()
+                   << "\t++numTransposeEliminated = " << numTransposeEliminated
+                   << " due to all the users of curOp are all transposed\n");
+        if (isa<mhlo::TransposeOp>(curOp)) {
+          SmallVector<int64_t> operandPerm;
+          SmallVector<int64_t> newPermutation;
+          if (failed(backwardTransposePermutation(curOp, permMap[curOp],
+                                                  operandPerm, newPermutation)))
+            return failure();
+          if (isIdentityPermutation(newPermutation)) {
+            ++numTransposeEliminated;
+            LLVM_DEBUG(llvm::dbgs() << "\t++numTransposeEliminated = "
+                                    << numTransposeEliminated
+                                    << " due to curOp is transpose op and can "
+                                       "be eliminated after folding\n");
+          }
+        }
+      }
+      LLVM_DEBUG(llvm::dbgs() << "\tpropagate transpose to operands of curOp ("
+                              << curOp->getNumOperands() << ")\n");
+      for (const auto& en : llvm::enumerate(curOp->getOperands())) {
+        LLVM_DEBUG(llvm::dbgs() << "\t\tcheck operand #" << en.index()
+                                << " with numTransposeEliminated = "
+                                << numTransposeEliminated << "\n");
+        auto& operandPerm = operandPermMap[en.index()];
+        if (isIdentityPermutation(operandPerm)) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "\t\tskip propagate for operand #" << en.index()
+                     << " due to operandPerm is identity\n");
+          continue;
+        }
+        auto operandDefiningOp = en.value().getDefiningOp();
+        // transpose can be folded for const op.
+        if (dyn_cast_or_null<mhlo::ConstantOp>(operandDefiningOp)) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "\t\tskip propagate for operand #" << en.index()
+                     << " due to operand op is const op\n");
+          continue;
+        }
+        bool needInsertTransposeOp =
+            tryUpdate(operandDefiningOp, curOp, operandPerm);
+        if (needInsertTransposeOp) {
```

Comment:
so the current 'cost' is number of transpose op? Not the data amount transposed ?

Reply:
Anyway, let's start here. Refine later, maybe.